qiskit/transpiler/passes/graysynth.py

```diff
@@ -0,0 +1,250 @@
+    This function is an implementation of the GraySynth algorithm.
+    The algorithm is described in the following paper in section 4:
+    "On the controlled-NOT complexity of controlled-NOT–phase circuits."
+    Amy, Matthew, Parsiad Azimzadeh, and Michele Mosca.
+    Quantum Science and Technology 4.1 (2018): 015002.
+
+    Args:
+        cnots (list): as described in the aforementioned paper.
+        number (int): the number of quantum bits in the circuit
+        nsections (int): the number of sections
+
+    Returns:
+        QuantumCircuit: the quantum circuit
+    """
+
+    # Create a Quantum Register with n quantum bits.
+    qreg = QuantumRegister(number, 'q')
+    # Create a Quantum Circuit acting on the q register
+    qcir = QuantumCircuit(qreg)
+
+    range_list = list(range(number))
+    epsilon = number
+    sta = Storage()
+    cnots_copy = np.transpose(np.array(copy.deepcopy(cnots)))
+    state = np.eye(number).astype('int')  # This matrix keeps track of the state in the algorithm
+
+    # Check if some T-gates can be applied, before adding any C-NOT gates
+    for qubit in range(number):
+        index = 0
+        for icnots in cnots_copy:
+            if np.array_equal(icnots, state[qubit]):
+                qcir.t(qreg[qubit])
+                cnots_copy = np.delete(cnots_copy, index, axis=0)
+                if index == len(cnots_copy):
+                    break
+                index -= 1
+            index += 1
+
+    # Implementation of the pseudo-code (Algorithm 1) in the aforementioned paper
+    sta.push([cnots, range_list, epsilon])
+    while not sta.isempty():
+        [cnots, ilist, qubit] = sta.pop()
+        if cnots == []:
+            continue
+        elif 0 <= qubit < number:
+            condition = True
+            while condition:
+                condition = False
+                for j in range(number):
+                    if (j != qubit) and (sum(cnots[j]) == len(cnots[j])):
+                        condition = True
+                        qcir.cx(qreg[j], qreg[qubit])
+                        state[qubit] ^= state[j]
+                        index = 0
+                        for icnots in cnots_copy:
+                            if np.array_equal(icnots, state[qubit]):
+                                qcir.t(qreg[qubit])
+                                cnots_copy = np.delete(cnots_copy, index, axis=0)
+                                if index == len(cnots_copy):
+                                    break
+                                index -= 1
+                            index += 1
+                        for x in remove_duplicates(sta.elements() + [[cnots, ilist, qubit]]):
+                            [cnotsp, _, _] = x
+                            if cnotsp == []:
+                                continue
+                            for ttt in range(len(cnotsp[j])):
+                                cnotsp[j][ttt] ^= cnotsp[qubit][ttt]
+        if ilist == []:
+            continue
```

Comment:
in the original pseudocode this check happens at the same place as 
```
        if cnots == []:
            continue
```
is there a reason you have separated them here?

Reply:
Yes, there is a reason for separating those two arguments. This is actually a mistake in the pseudocode, and it took me quite a while to find this out. If they were in the same line (as in the pseudocode, most of the CNOT's are actually missed). 

Here is another implementation of the GraySynth algorithm, written by my colleague:
https://github.com/boschmitt/tweedledum/blob/master/include/tweedledum/algorithms/synthesis/gray_synth.hpp

He also separated those two conditions.